### ISSUE TITLE
Text overflow with ellipsis is weird and ugly by design

### ISSUE THREAD

**Issue by marcglasberg** [flutter/flutter#18761]
_Latest update: https://github.com/flutter/flutter/issues/18761#issuecomment-1404207970_

----

Internal: b/137284394

Here (https://github.com/flutter/flutter/issues/4478) Hixie commented:

   *"Ok the plan is that rather than truncating individual characters until it fit, we will truncate entire atomic positioned glyph sequences (e.g. whole ligatures or shaped words) until the ellipsis fits. That would have the advantage of being much more performant. It would have the disadvantage of truncating the entire word if the whole line was one ligature (e.g. if your line was the word "Zapfino" in the Zapfino font, and it overflowed at the "o", we would replace the whole thing with a "...", rather than just the "no".)"*

Although the above comment talks about ligatures, it seems to me this is being ignored, and whole words are simply being removed.

Now suppose the text `"Laboris nisi ut aliquip ex ea commodo"` gets cut right before word `"ex"`. Since the previous word fits, it may cut some letters of that last word to fit the ellipsis, resulting in:

`"Laboris nisi ut aliqu..."`

However, if the horizontal space is a litter smaller, than the letter `"p"` from `"aliquip"` won't fit, and it will cut the whole word, resulting in:

`"Laboris nisi ut ...          "`

But that's not how overflow works for other platforms, and it makes no sense to remove whole words. The obvious result, if "p" doesn't fit should be, just as before:

`"Laboris nisi ut aliqu..."`

As things are now, we tend to see a lot of words cut in their last 1 or 2 chars, and a lot of words completely removed, leaving a lot of weird white space. That gets particularly nasty in one line texts like titles.

Please, don't create weird behaviors for things with time-honored traditions like ellipsizing. Stick to what everybody else does. If you want, let the devs choose between both behaviors: `enum TestOverflow {clip, fade, ellipsisWords, ellipsisLetters}`.

If Arabic and other eastern languages have problems with that (as per issue 4478), you can even do some autodetection for characters of those languages, by adding the enum `TestOverflow.ellipsisAutodetect`.

**Comment by sir-boformer** [comment]
I agree with you, the current ellipsis often looks out of place.

**Comment by HannoOttens** [comment]
This is an issue for me aswell.
I have some right-aligned text that frequently overflows, resulting in a large white gap between the ellepsis and the right side of screen. This looks _very_ odd.

**Comment by josvos** [comment]
+1 here... ellipsis on just word boundaries and leaving (much) space unused after the ellipsis is not the right way, IMHO, at least not for small texts (e.g. labels in buttons, etc.).  Please change this, optionally by adding `ellipisChars` or so to let devs choose (not needed for me, but if the word boundary is needed too).

**Comment by ConProgramming** [comment]
Yeah I have an email in A Text widget and the entire email gets cut out

**Comment by sir-boformer** [comment]
You can try this as a workaround. Not sure if it also works with ellipsis.
```dart
      Text(
        email,
        maxLines: 1,
        overflow: TextOverflow.fade,
        softWrap: false,
      );
```

**Comment by JonasJW** [comment]
@sir-boformer sadly `TextOverflow.ellipsis` doesn't work like this but that is the way it should work in my opinion. Just replace the letters which are fading with '...'.

**Comment by Hixie** [comment]
cc @GaryQian

**Comment by lukepighetti** [comment]
`TextOverflow.ellipsis` should work on vertical overflow, cut off the last word(s) and replace with `...`

**Comment by 1343** [comment]
Some issue I am also facing, if some word is fitting it's replacing the complete word 
**Text(   'Sachin Ramesh Tendulkar ',
                    softWrap: false,
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                    style: new TextStyle(
                      fontSize: 13.0,
                      fontFamily: 'Roboto',
                      color: new Color(0xFF212121),
                      fontWeight: FontWeight.bold,
                    ),
                  )**
If **er** is not wraping it's removing complete tendulkar word and putting ... final text **Sachin ramesh ..**

**Comment by chunhtai** [comment]
I also saw some weird behavior with the maxLines and overflow.
it seems like maxLines is ignored the if the overflow is set.

**Comment by chunhtai** [comment]
Current ellipsis is based on soft line break that automatically inserted in paragraph builder. When the word can't fit in to the line, it will get pushed to next line, and that cause it to truncate the whole word if part of it cannot fit. I will come up with a new overflow enum that will truncate the line differently.

**Comment by Moncader** [comment]
@chunhtai 
Been a while since an update.
Any word on progress?

In languages where there are no spaces built in to the language like Japanese, this problem is extremely visible when you mix an alphanumeric sequence with a Japanese string of characters.
This is extremely common for business names, project names, or movie titles, which is exactly the kind of string you don't want to cut as much.

**Comment by duzenko** [comment]
How does it work in regards to hyphens? E.g. I have a string 'Franc Krona - Food stand - #09-12-2019-#011'
It seems to cut at a spaced hyphen leaving huge empty space while I would actually like it to cut an any hyphen, spaced or not.

**Comment by vanlooverenkoen** [comment]
Any progress on this issue?

**Comment by chunhtai** [comment]
@jason-simmons Do you know what is the status of txt skia migration?

**Comment by sparcsaswat** [comment]
DropdownMenuItem(
                          value: value.r_value,
                          child: Text(
                            value.r_text,
                            overflow: TextOverflow.fade,
                            softWrap: true,
                          ),
                        )

In my case it is not working

**Comment by JagandeepBrar** [comment]
Any updates at all?

@chunhtai mentioned 1-2 months at the beginning of September and we are getting close to 3 months now.

**Comment by Reprevise** [comment]
To demonstrate the _awkward_ space that a TextOverflow brings, here:

There's a large amount of text here and with an overflow, it leaves a huge space as @HannoOttens stated. In this case, I'm using a `Column` inside of a `Row`. Surrounding the `Column` is a `Container` that has a set width and its parent is a `Flexible` widget which itself is located in the aforementioned `Row`. With `softWrap` enabled for both of the `Text` widgets and both of them using `TextOverflow.fade`.
![flutter text overflow issue](https://user-images.githubusercontent.com/25782944/69288633-23c99680-0bc8-11ea-982f-bbb64ec91b84.png)

Sidenote: How will languages like Arabic be affected by text overflows? Their characters can't just be cut off like English ones.

**Comment by FabianGuerreroMaureira** [comment]
Any update?

**Comment by marcglasberg** [comment]
Hello everyone, I'm publishing a workaround to this issue (for a single line at least) in this package: 

   https://pub.dev/packages/assorted_layout_widgets

## TextOneLine

`TextOneLine` is a substitute for `Text` when `maxLines: 1`. It uses a special fade-with-ellipsis, which is
much better than the current buggy and ugly-looking ellipsis-that-cuts-the-whole-word.

For example, this:

```dart
Text("This isAVeryLongWordToDemonstrateAProblem", maxLines: 1, softWrap: false);  
```

Will print this in the screen:
```
This ...  
```

While this:

```dart
TextOneLine("This isAVeryLongWordToDemonstrateAProblem");  
```

Will print this:
```
This isAVeryLongWordToDemonst...  
```

I created this widget a long time ago, right after I filled this issue, but never published it. Since the new Flutter version is now out without a fix, I decided to publish it now. To maintain it is a pain, since it breaks every time Flutter upgrades. So I still hope this issue is fixed and I can deprecate this widget once and for all. 

Maybe one of you people could do a PR to Flutter with this code?... I'd suggest `TextOverflow` could have an extra value `TextOverflow.fadeWithEllipsis` that adds this effect.

**Comment by benPesso** [comment]
@marcglasberg Would be nice if you had a workaround for multi-line text as well.

**Comment by chunhtai** [comment]
This is currently blocked on skia migration. https://github.com/flutter/flutter/issues/39420.

**Comment by marcglasberg** [comment]
@benPesso Sure it would, but I don't have it, because I don't need it in my project.

@chunhtai Any ETA?

**Comment by chunhtai** [comment]
@marcglasberg we currently still do not have ETA because the migration is causing a lot of test breakages due to small pixel to pixel change. I will update this issue as soon as we have a target date

**Comment by feinstein** [comment]
I just want to add my current issue with this to illustrate how bad this is.

I have a button that should say:

![image](https://user-images.githubusercontent.com/8120472/71288775-2dbaf280-234a-11ea-866f-c966ef10692d.png)

But if your email is too long, it just shows:

![image](https://user-images.githubusercontent.com/8120472/71288800-43301c80-234a-11ea-9481-39d929ce9260.png)

Which is completely frustrating, as not only it lacks any information to the user, it also wastes all that space at the right, as if the text is there somehow taking up the space in that line, forcing the remainder text to the left.

Thanks @marcglasberg for your workaround package.

**Comment by marcglasberg** [comment]
@feinstein Put the `TextOneLine` inside of a `FitHorizontally` widget of that same `assorted_layout_widgets` package, and you'll gain 50% more space before it even has to use ellipsis.

**Comment by YAPPDEF** [comment]
Any update on this?

**Comment by ldelafuente** [comment]
It's been more than a year and a half since the issue was opened. Do you need help in any way so we can get a fix for this?

**Comment by chunhtai** [comment]
There is still no update at this point.

**Comment by Reprevise** [comment]
It's been a month and a half, any update?

**Comment by chunhtai** [comment]
@zoeyfan can you provide some insight on why this become customer critical. The current plan is still waiting for the skia migration i mentioned in https://github.com/flutter/flutter/issues/18761#issuecomment-567599627

**Comment by zoeyfan** [comment]
@chunhtai, this was brought to my attention my one of our top customers and they indicated that this issue was quite urgent on their side. Can we follow up with Skia to get an ETA from them?

**Comment by chunhtai** [comment]
@zoeyfan the migration is taken care by the engine team. Can you also update the tracking issue https://github.com/flutter/flutter/issues/39420 to be custom critical and we can continue the discussion there?

**Comment by zoeyfan** [comment]
Connected with the customer (BD), and suggested to turn on --enable-skshaper flag to see if the issue can be resolved. Customer is verifying the workaround, and will respond once they have progress to report.

**Comment by rodi189** [comment]
Hello anybody have news about this issue? i have this problem with words like "AB CDEFGHIJK LMNO" is showing like "AB..." is annoying.

**Comment by Hixie** [comment]
We're going to lower the priority to P3 since we haven't heard back from our customers on this one despite reaching out. Please let us know if it is blocking you! The reality is that this is something we're unlikely to fix until after we've done the SkParagraph migration (#39420), which probably won't be any time soon since we want to make sure we do it carefully without regressing anything. Thanks for your understanding.

**Comment by benPesso** [comment]
Hey @Hixie. This isssue is a big pain for many developers, not just the reporter. We've all been waiting for this to be resolved, and have resorted to ugly workarounds in our apps because of this.

If anything, priority should go up, not down.

**Comment by marcglasberg** [comment]
@benPesso I agree the priority should go up, but for the moment, at least for one-line texts, you can use my **TextOneLike** widget from https://pub.dev/packages/assorted_layout_widgets

**Comment by Hixie** [comment]
Don't get me wrong, we agree it's important. According to our [priority levels](https://github.com/flutter/flutter/wiki/Issue-hygiene#priorities), P3 is the highest priority level that isn't in the top 30 or so bugs. This is important, but I hope you agree that it isn't a top-30 bug. :-)

**Comment by benPesso** [comment]
@marcglasberg Yeah, I'm familiar with your widget. It's a great workaround (and I almost went through the trouble of modifying it to work with two-line text) but in the end I decided it was better to change the UI than use a workaround.

@Hixie I completely understand. Though, as far as I'm concerned, there are no other bugs that are more important. (Well, except for #55061, maybe. 😄)

**Comment by pph7** [comment]
@Hixie this is still a very important bug, it should have the highest possible priority. Text rendering is ugly in many cases because of that bug. Especially in languages with longer words. Hope you reconsider the P3 placement. Thank you!

**Comment by Hixie** [comment]
The highest possible priority (P0) is what we reserve for bugs like "The Flutter framework won't compile" or "All Flutter apps crash when installed from the play store" or "We have regressed the performance of all Flutter apps by 50%". The next two levels (P1 and P2) are for bugs that a contributor has to work on because their employer depends on it. This bug's priority (P3) is the highest possible priority for all other kinds of bugs.

**Comment by Pe-te** [comment]
`.replaceAll("", "\u{200B}")`

muahahaha! 😜

no guarantee if it works with weird languages.

**Comment by Eerey** [comment]
> .replaceAll("", "\u{200B}")

thank you. this is actually a good workaround

**Comment by Mervin1024** [comment]
> `.replaceAll("", "\u{200B}")`
> 
> muahahaha! 😜
> 
> no guarantee if it works with weird languages or multiline.

Thanks，It's worked for me.

**Comment by strifel** [comment]
Worked for me, too. (Even in german)
Thanks.

**Comment by carsguide-admin** [comment]
> .replaceAll("", "\u{200B}")

What is the magic???? 😜 THANK YOU

**Comment by kkizlaitis** [comment]
> `.replaceAll("", "\u{200B}")`
> 
> muahahaha! 😜
> 
> no guarantee if it works with weird languages or multiline.

Unfortunately, that does not work with multiline text, when `maxLines` is not defined. 😞

**Comment by derChris77** [comment]
.replaceAll("", "\u{200B}") works fine for maxLines = 1 but when there are more lines the line breaks are in the middle of a word which gets really ugly...
Is there a solution for this problem??

**Comment by pedromassangocode** [comment]
Still valid on latest master channel.

<details><summary>sample code</summary>

```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
          appBar: AppBar(title: Text('Test'),),
          body:Container(
            padding: EdgeInsets.all(20),
            child: Column(
              children: [
                Container(
                    color: Colors.red,
                    child: Text('哈哈134567891234561124444123456112441234561451241345',maxLines: 1,overflow: TextOverflow.ellipsis)),
                SizedBox(height: 10,),
                Container(
                    color: Colors.teal,
                    child: Text('word134567891234561124444123456112441234561451241345',maxLines: 1,overflow: TextOverflow.ellipsis)),
              ],
            ),
          )

      ),
    );
  }
}
```
</details>

<details><summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.87, on Mac OS X 10.15.6 19G2021, locale en)
    • Flutter version 1.22.0-10.0.pre.87 at /Users/pedromassango/dev/SDKs/flutter_master
    • Framework revision 4732a214a7 (3 days ago), 2020-09-04 21:05:02 -0400
    • Engine revision ac8b9c4c52
    • Dart version 2.10.0 (build 2.10.0-86.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.7)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.7, Build version 11E801a
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 48.1.4
    • Dart plugin version 202.7206

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (4 available)
    • iPhone 11 Pro Max (mobile) • 3DA251B5-C982-48C8-8FF5-C4D1DB21ED89 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-13-7 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • Mac OS X 10.15.6 19G2021
    • Web Server (web)           • web-server                           • web-javascript • Flutter Tools
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 85.0.4183.83

! Doctor found issues in 1 category.

```
</details>

**Comment by yuhaocan** [comment]
> `.replaceAll("", "\u{200B}")`
> 
> muahahaha! 😜
> 
> no guarantee if it works with weird languages or multiline.

But my text is "橙美童🍊" will crash

**Comment by Oklahomawhore** [comment]
Please fix ASAP. Thank you.  It is an unbelievable miss.

**Comment by yuhaocan** [comment]
Finally I use regular expression  to deal  problem. But I hope Flutter would fix this issue.
```Dart
 String message = "橙美童kkkk🍊";
 message = "哈哈哈ssssddfgggllllllllkjjljljkl";
 final exp = new RegExp('[\u4e00-\u9fa5]');
 String text = message.replaceAllMapped(exp,
            (Match m) => "${m[0]}\u2060");
```

**Comment by yuhaocan** [comment]
> Finally I use regular expression to deal problem. But I hope Flutter would fix this issue.
> 
> ```dart
>  String message = "橙美童kkkk🍊";
>  message = "哈哈哈ssssddfgggllllllllkjjljljkl";
>  final exp = new RegExp('[\u4e00-\u9fa5]');
>  String text = message.replaceAllMapped(exp,
>             (Match m) => "${m[0]}\u2060");
> ```

But if String is "哈🀂🀁🀄︎🀖哈🀂🀁🀄︎🀖哈ssssddfgggllllllllkjjljljkl🀂🀁🀄︎🀖🀠＠@㏒∮ΡΟㄇㄈㄞㄞㄝ" still failed

**Comment by PiN73** [comment]
Just noticed that Flutter actually _does_ put ellipsis in the middle of word sometimes (blue "v"s).

But sometimes it doesn't while definitely could (red "x"s).

I can't figure out what it depends on. Flutter 1.20.4, [source code](https://github.com/PiN73/test_ellipsis/blob/master/lib/main.dart)

![screenshot](https://user-images.githubusercontent.com/18294052/93946935-37431200-fd43-11ea-99fb-f9b99df30834.png)

**Comment by marcglasberg** [comment]
@PiN73 To solve your mystery, It put correct ellipsis when the word being cut would fit the available space, but there are more words. If the word being cut would not fit the available space, it cuts the whole word out.

**Comment by marcglasberg** [comment]
Also I'd say this doesn't work because char 32 (space) can be a part of an Unicode char that takes more than one byte.:

```dart
.replaceAll(" ", "\u{200B}")
```

I'd suggest two solutions:

1) Once again, my `TextOneLine` widget from https://pub.dev/packages/assorted_layout_widgets

2) Using the Characters package (https://pub.dev/packages/characters). I haven't tested, but something along these lines should work:

```dart
var text = "My original text";
var safeText = Characters(text).replaceAll(Characters(" "), Characters("\u{200B}").toString();

// And the use it:
Text(safeText);
```

This will correctly remove spaces and add non-break spaces. 

However, both solutions only work for fixing 1 line texts.

**Comment by yuhaocan** [comment]
> Also I'd say this doesn't work because char 32 (space) can be a part of an Unicode char that takes more than one byte.:
> 
> ```dart
> .replaceAll(" ", "\u{200B}")
> ```
> 
> Also this doesn't work because you'd be adding non-break spaces between chars of an Unicode char that takes more than one byte:
> 
> ```dart
> .replaceAll("", "\u{200B}")
> ```
> 
> I'd suggest 3 solutions:
> 
> 1. Once again, my `TextOneLine` widget from https://pub.dev/packages/assorted_layout_widgets
> 2. Using the Characters package (https://pub.dev/packages/characters), you can remove spaces and add non-break spaces. I haven't tested, but something along these lines should work:
>    ```dart
>    var text = "My original text";
>    var safeText = Characters(text).replaceAll(Characters(" "), Characters("\u{200B}").toString();
>    
>    // And the use it:
>    Text(safeText);
>    ```
> 3. Using the Characters package again to add non-break spaces between REAL chars (respecting Unicode chars that take up multiple bytes). Something along the lines of (not tested):
>    ```dart
>    var text = "My original text";
>    var safeText = Characters(text).toList().join("\u{200B}");
>    
>    // And the use it:
>    Text(safeText);
>    ```
> 
> However, all these solutions only work for fixing 1 line texts.

TextOneLine cant solve the fundamental problem. In UI effect we only need ellipsis no TextOverflow.fade.

**Comment by marcglasberg** [comment]
@yuhaocan `TextOneLine` and the `Characters` solutions are just workarounds to help us wait until the real solution comes from the Flutter team. None of these workarounds work for multiline anyway.

**Comment by acoutts** [comment]
I find it's really hard to make text behave correctly with overflows while in one line. Hoping for a fix soon.

**Comment by marcglasberg** [comment]
Please vote/give your opinion about a change in Flutter's non-breaking policy: https://github.com/flutter/flutter/issues/70613

If this policy I propose were in place when they made the change to the text overflow, we wouldn't have had this ellipsis problem to start with.

**Comment by pushpendraKh** [comment]
Any update on this? Is this going to be resolved or we need to apply some hack from our code side only? 
cc: @chunhtai

**Comment by wei120698598** [comment]
The problem fixed?Any wold process?did you  test  The words 呵呵哈哈123456

**Comment by SilkeNL** [comment]
the `.replaceAll("", "\u{200B}")` fix worked for me with the code:
`Text(
                  textString.replaceAll("", "\u{200B}"),
                  style: TextStyle(color: Colors.grey[600]),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  softWrap: false,
                )`

**Comment by The-Redhat** [comment]
> the `.replaceAll("", "\u{200B}")` fix worked for me with the code:
> `Text( textString.replaceAll("", "\u{200B}"), style: TextStyle(color: Colors.grey[600]), maxLines: 1, overflow: TextOverflow.ellipsis, softWrap: false, )`

You have to be careful with this solution because it breaks when you use emojis or other special characters. You should use the `characters` package instead.

I created an extension in my project
```
extension StringX on String {
  String get overflow => Characters(this)
      .replaceAll(Characters(''), Characters('\u{200B}'))
      .toString();
}
```

Which then can be used the following way:
```
Text(
     textString.overflow,
     overflow: TextOverflow.ellipsis,
     maxLines: 1,
)
```

**Comment by marcglasberg** [comment]
This is the current code I am currently using fot the `TextOneLine` widget in the https://pub.dev/packages/assorted_layout_widgets:

```
class TextOneLine extends Text {
  const TextOneLine(
    String data, {
    Key? key,
    TextStyle? style,
    StrutStyle? strutStyle,
    TextAlign? textAlign,
    TextDirection? textDirection,
    Locale? locale,
    TextOverflow? overflow = TextOverflow.ellipsis,
    double? textScaleFactor,
    String? semanticsLabel,
    TextWidthBasis textWidthBasis = TextWidthBasis.parent,
    TextHeightBehavior? textHeightBehavior,
  }) : super(
          data,
          key: key,
          style: style,
          strutStyle: strutStyle,
          textAlign: textAlign,
          textDirection: textDirection,
          locale: locale,
          softWrap: false,
          overflow: overflow,
          textScaleFactor: textScaleFactor,
          maxLines: 1,
          semanticsLabel: semanticsLabel,
          textWidthBasis: textWidthBasis,
          textHeightBehavior: textHeightBehavior,
        );

  @override
  String? get data => super.data == null ? null : Characters(super.data!).toList().join("\u{200B}");
}
```

**Comment by nxcco** [comment]
This little feature was requested in 2018 and there is still no real progress? What's the problem?

**Comment by dawid-niedzwiecki** [comment]
Come on, why is this still open?

**Comment by kira2015** [comment]
The problem still exists!

**Comment by pph7** [comment]
Indeed it does.. More than 3 years later. I mean, I love flutter and the development team is doing a great job. But I am starting to think they don't care much about any languages than english.

**Comment by RaymondBakker** [comment]
Curious to see this is still not resolved.

**Comment by chunhtai** [comment]
last time i check we still do not fully opt in to skia https://github.com/flutter/engine/blob/90424212a3d6839ca7552b84f6d1108042bd0d94/common/config.gni#L18

This feature requires changes to skia typographic API.

**Comment by feinstein** [comment]
@chunhtai should we open an issue about that API change? Or is it open already and we just need to upvote it?

**Comment by chunhtai** [comment]
This is the skia migration https://github.com/flutter/flutter/issues/39420, which is also pretty old too. I am not entirely sure what is the current progress on that issue neither. We may not get an update until people come back from vacation.

Once that is done, we can use this issue to track the progress of API change.

**Comment by ventr1x** [comment]
This is attrocious, like so much from the Flutter team.
fade is NOT an alternative, my clients constantly complain about this and EVERYONE finds fade out weird and actually thinks it is a bug. Ellipsis is not working like anyone ever woud expect. Cutting off entire sentences because they might not fit? Makes no sense and has never been done that way.
But we all know this will never be fixed. Nothing new at all.

**Comment by tvolkert** [comment]
> But we all know this will never be fixed. Nothing new at all.

@ventr1x per our [code of conduct](https://github.com/flutter/flutter/blob/master/CODE_OF_CONDUCT.md), we ask that people "Be kind. Be courteous. Be welcoming."  Moreover, you can see from the comment immediately preceding yours that this issue is blocked by another issue that is actively being worked on.

**Comment by duzenko** [comment]
You can't reasonably expect to ignore community needs for years and not get any lashback, can you?

**Comment by ventr1x** [comment]
Besides that this should've been implemented correctly in the first place, "actively being worked on" does not really sit well after 4 years "actively being worked on". 
Such issues, that people often find too late in their development and clients refusing payments because of it, are quite common with Flutter. A simple recent example are the emoji issues, also "actively being worked on" for years. They might seem smaller issues but are very disruptive and time consuming for hundreds of devs.
As this is often an underlying architectural issue because the Flutter team did not have the proper education in the beginning (your Lead Engineer confirmed this on reddit), we can't even contribute to it properly.

**Comment by Kavantix** [comment]
@chunhtai is there anything we can do to help?
The issue you linked seems to not have had any actual activity since november 2020, but it sounds like this is pretty much implemented and just needs testing
So perhaps we can help test?

**Comment by chunhtai** [comment]
@Kavantix To test the skparagaph, you need to build custom engine with additional flag.

```
flutter/tools/gn. .... --enable_skshaper --always_use_skshaper
```

There is a PR to that should switch to use Skia paragraph by default https://github.com/flutter/engine/pull/28912/files
Once that land, we should be able to start progress on this issue

**Comment by Kavantix** [comment]
I’ve done some testing using the skia pragraph.
So far it seems to behave exactly the same as the old paragraph implemention.
Including the weird behavior when using the ellipsis as discussed in this issue.

So I’ve also done some digging where the root cause of this is, and as was said before it is indeed because the long word is moved to the next line and the ellipsis is simply added because it is exceeding the maximum number of lines.
All of this is determined inside skparagraph, so to resolve this issue, it needs to be changed inside skia.

One way of fixing the issue would be to never break the last line when the maximum is reached.
I tested simply handling the last line as one that does not allow soft breaks with a small change in skia and it did keep the word but not yet the ellipsis.

**Comment by chunhtai** [comment]
The problem is there is no API in typographic library (both skia and txt) that we used to calculate layout to set soft line break strategy. They always insert soft line break at words boundaries. No matter how we fix it, either adding additional API to change the line break strategy or moving the word around after they finishes layout, the fix is on the library side.

**Comment by chunhtai** [comment]
> So I’ve also done some digging where the root cause of this is, and as was said before it is indeed because the long word is moved to the next line and the ellipsis is simply added because it is exceeding the maximum number of lines. All of this is determined inside skparagraph, so to resolve this issue, it needs to be changed inside skia.

That is exactly the reason why we need to way for the migration, since the fix will be on the library side

**Comment by chunhtai** [comment]
Since this is unblocked, I am curious about how community feels about the current behavior. Is there a use case where the current behavior is preferred? Should there be an option to toggle between the proposed behavior and the existing behavior

**Comment by jsroest** [comment]
I would just introduce an ellipsis2, just not to break things.

**Comment by benPesso** [comment]
> Since this is unblocked, I am curious about how community feel about the current behavior. Is there a use case where the current behavior is preferred? Should there be an option to toggle between the proposed behavior and the existing behavior

@chunhtai Personally, I don't see why anyone would find the current implementation preferable. It's an issue that has been outstanding since day one of Flutter, and probably no one is happy with the way it currently works. But if we're worried about breaking stuff, I recommend adding a `clip` property (similar to the [`softWrap` property](https://api.flutter.dev/flutter/widgets/Text/softWrap.html), but an `enum`) which would allow the user to set the clipping mode of the text (`syllable`, `word`, `linebreak?`):
```dart
Text(
  'Some text',
  clip: TextClip.syllable,
)
```

---

> I would just introduce an ellipsis2, just not to break things.

@jsroest The ellipses is about how the clipped text should be indicated (visualized); not about the clipping behavior. So there's no need for a second `ellipses` property.

**Comment by ookami-kb** [comment]
> But if we're worried about breaking stuff, I recommend adding a clip property (similar to the [softWrap property](https://api.flutter.dev/flutter/widgets/Text/softWrap.html), but an enum) which would allow the user to set the clipping mode of the text (syllable, word, linebreak?)

If it would default to `syllable`, it would still be a breaking change. If it would default to a `word`, it means that everybody who thinks this is a bug (including me) will have to add an explicit clip option to every `Text` widget.

**Comment by marcglasberg** [comment]
@chunhtai Please see the original issue. It seems Arabic and other eastern languages have use for the current behavior. I have suggested the current behavior could be kept under one of the enums, like so:

```
enum TestOverflow {clip, fade, ellipsisWords, ellipsisLetters, ellipsisAutodetect}.
```

**Comment by Pe-te** [comment]
> If it would default to `syllable`, it would still be a breaking change. If it would default to a `word`, it means that everybody who thinks this is a bug (including me) will have to add an explicit clip option to every `Text` widget.

I would consider the current state "broken" and the new state "fixed", so it's not really a breaking change but a fixing one? :)

Maybe there is some way to make the option high level? So whoever needs it can define "all textfields in my project need the old way" with a simple toggle?

**Comment by ookami-kb** [comment]
> I would consider the current state "broken" and the new state "fixed", so it's not really a breaking change but a fixing one? :)

Yes, exactly my point of view as well.

> Maybe there is some way to make the option high level? So whoever needs it can define "all textfields in my project need the old way" with a simple toggle?

Adding it to `TextStyle`, probably?

**Comment by benPesso** [comment]
> @chunhtai Please see the original issue. It seems Arabic and other eastern languages have use for the current behavior. I have suggested the current behavior could be kept, like so:
> 
> ```
> enum TestOverflow {clip, fade, ellipsisWords, ellipsisLetters, ellipsisAutodetect}.
> ```

Arabic and other languages with varying glyph sizes at the end don't "need" the current behavior. They need it to properly  detect what the final width of the word is going to be, and do more iterations of the clipping logic, to determine where to correctly trim the word.
Clipping a word in the middle should not convert the (now final) glyph to a final one. In Hebrew, for example, you might have this word: `נכון`. Clipped in the middle, it would turn into `נכ`, not `נך`. The same should be true in Arabic and languages where glyphs are comprised of more than one letter - the singular version of the letter should be used.

**Comment by benPesso** [comment]
> > But if we're worried about breaking stuff, I recommend adding a clip property (similar to the [softWrap property](https://api.flutter.dev/flutter/widgets/Text/softWrap.html), but an enum) which would allow the user to set the clipping mode of the text (syllable, word, linebreak?)
> 
> If it would default to `syllable`, it would still be a breaking change. If it would default to a `word`, it means that everybody who thinks this is a bug (including me) will have to add an explicit clip option to every `Text` widget.

There's no argument that this is a bug. If anything, this solves just one bug, and uncovers another: the proper clipping of script-based languages. So I would consider this a bug fix.

However, since Flutter already has native support for locales, I don't see why this can't be handled internally based on the locale used for the string. It shouldn't be something the user has to configure (override) manually.

That said, I do suggest adding the `clip` property for those who wish to use this functionality in locales that don't natively require it. This way, locales that have it "on" by default (English, etc.) can use `TextClip.word` to achieve the old behavior, and locales that don't (Arabic, etc.) can use the `TextClip.syllable` property.

**Comment by benPesso** [comment]
> > I would consider the current state "broken" and the new state "fixed", so it's not really a breaking change but a fixing one? :)
> 
> Yes, exactly my point of view as well.
> 
> > Maybe there is some way to make the option high level? So whoever needs it can define "all textfields in my project need the old way" with a simple toggle?
> 
> Adding it to `TextStyle`, probably?

`TextStyle` is for styling, not formatting. I remember seeing a discussion a long while back about why stylings are grouped into a`TextStyle` and formattings are not, but that's a discussion for another day. 😄

**Comment by marcglasberg** [comment]
I mentioned Arabic because that was the justification for the current behavior, here: https://github.com/flutter/flutter/issues/4478

If that makes sense or not, I can't judge.

**Comment by davidsdearaujo** [comment]
Hey, look this workaround. Maybe it can help to solve hide "letter by letter" on overflow before the correct solution:

## Steps to do
1. Create a `Text` to each letter;
2. Include all the `Text` in a `Row`;
3. Get the `Row` rect;
4. Get each `Text` rect;
5. Validate if the `Text` rect is in the `Row` rect and, case isn't in the rect, hide it;

## I found this problems
1. The character spacing is little bigger comparing with simple `Text` widget;
2. The `Row` is duplicated to get the row size after the first frame is rendered;

I tried to implement it in this package: [text_helpers](https://pub.dev/packages/text_helpers)

**Comment by davidsdearaujo** [comment]
> That's a terribly inefficient way of doing things. Just have a look at the [underlying code for the `Text` widget](https://github.com/flutter/flutter/blob/2f73173c36f806c7f30e705a90722082f85a2c8b/packages/flutter/lib/src/widgets/text.dart#L520), and you'll see why.

Thank you for your feedback.

Yes, terribly inefficient way, but it works.
I'm waiting for an oficial solution or another that works and is more efficient.

In the meantime, I'll try to make a more efficient solution.

**Comment by MatrixDev** [comment]
@davidsdearaujo , if you want at least somewhat more efficient solution, look here - https://itnext.io/writing-custom-widgets-in-flutter-part-1-ellipsizedtext-a0efdc1368a8. Using TextPainter is much more efficient that widget-per-character.

**Comment by m-schmailzl** [comment]
Instead of using a Text-Widget for every letter, I am using this workaround for quite a while:

`Characters('your string').replaceAll(Characters(''), Characters('\u{200B}')).toString();`

You need the [characters](https://pub.dev/packages/characters) library.

I am using this approach in production for half a year and didn't have any problems. It even works with emojis.

**Comment by Linloir** [comment]
> Instead of using a Text-Widget for every letter, I am using this workaround for quite a while:
> 
> `Characters('your string').replaceAll(Characters(''), Characters('\u{200B}')).toString();`
> 
> You need the [characters](https://pub.dev/packages/characters) library.
> 
> I am using this approach in production for half a year and didn't have any problems. It even works with emojis.

Thank goodness this is what I'm looking for

Brilliant solution

**Comment by m-schmailzl** [comment]
> > Instead of using a Text-Widget for every letter, I am using this workaround for quite a while:
> > `Characters('your string').replaceAll(Characters(''), Characters('\u{200B}')).toString();`
> > You need the [characters](https://pub.dev/packages/characters) library.
> > I am using this approach in production for half a year and didn't have any problems. It even works with emojis.
> 
> Nice solution!
> 
> We can also use it without the `Character` class:
> 
> `'Lorem Ipsum is simply dummy text'.replaceAll('', '\u{200B}')`

This solution breaks when users are using emojis. That's why I use characters.

**Comment by davidsdearaujo** [comment]
> This solution breaks when users are using emojis. That's why I use characters.

Oh, understood.
The `character` package is already in Flutter. Now we have an extension to get the characters from `String`:
```dart
'Lorem Ipsum is simply dummy text'
    .characters
    .replaceAll(Characters(''), Characters('\u{200B}'))
    .toString()
```

Thanks you for your contribution!

**Comment by kamilpowalowski** [comment]
It is not a solution. It's a workaround. Be very cautious with using it. It changes how characters are displayed. Differences are subtle, but overall, letters are a bit wider.

**Comment by jaded0** [comment]
It's really weird that this is till a problem. I feel like the solution is a probably a simple one.

**Comment by thailocnexle** [comment]
> Characters('your string').replaceAll(Characters(''), Characters('\u{200B}')).toString()

wow, Nice solution, it helps me so much!

**Comment by SShayashi** [comment]
The workaround does not seem to work for multi-byte characters. I tried Chinese characters but the text has white space at the end. You can try the following sample code.

Some text does not have white space at the end (You can see it at the bottom of the text in the following sample). I believe this is the expected behavior.

Are there any solutions?

<details>
<summary>  Sample code </summary>

```dart
import 'package:assorted_layout_widgets/assorted_layout_widgets.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    const chineseCharacters = '漢字一漢字二漢字三漢字四漢字五漢字六漢字七漢字八';
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home Page'),
      ),
      body: Center(
        child: Container(
          width: 300,
          child: Column(
            children: [
              const Text(
                  'The following Chinese characters have white space at the end.'),
              Container(
                color: Colors.green,
                child: const Text(
                  chineseCharacters,
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
              const Text('The workaround does not work.'),
              Container(
                color: Colors.green,
                child: const TextOneLine(chineseCharacters),
              ),
              Container(
                color: Colors.green,
                child: Text(
                  chineseCharacters.characters.toList().join('\u{200B}'),
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
              Container(
                color: Colors.green,
                child: Text(
                  chineseCharacters.characters
                      .replaceAll(
                        Characters(''),
                        Characters('\u{200B}'),
                      )
                      .toString(),
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
              const Text('Some texts does not happen without the workaround.'),
              Container(
                color: Colors.green,
                child: const Text(
                  '漢字一 漢字二 漢字三 漢字四 漢字五 漢字六漢字七',
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```
</details>

Sample code result:
<img src="https://user-images.githubusercontent.com/12263798/168413387-c9c5655c-d176-4ef2-a4d4-10ed9c6f81e9.png" width="320">

**Comment by chunhtai** [comment]
Added upstream issue https://bugs.chromium.org/p/skia/issues/detail?id=13325

**Comment by mack-at-pieces** [comment]
+1 here. This workaround has proven to be the best, but agreed this is an inconvenient issue that has been around far too long.

`Text(
  'Text',
  style: Styles.font.caption(),
  maxLines: 1,
  overflow: TextOverflow.fade,
  softWrap: false,
),`

**Comment by shahbazhashmi** [comment]
Try this

https://gist.github.com/shahbazhashmi/6b1823922b587144ebb180a89af995a9

**Comment by roman-petrov** [comment]
> It is not a solution. It's a workaround. Be very cautious with using it. It changes how characters are displayed. Differences are subtle, but overall, letters are a bit wider.

Yes, this workaround increases letter spacing so it is not suitable for us. I just found another workaround which I not see in this thread:

```dart
 text.replaceAll(' ', '\u{000A0}')
```

We started to use it and it seems that it "just works" for all our use-cases. It just replaces all spaces in text with Unicode [No-Break Space](https://unicode-table.com/en/00A0/) character.

Hope it may help someone :)

**Comment by SShayashi** [comment]
> Yes, this workaround increases letter spacing so it is not suitable for us. I just found another workaround which I not see in this thread:
>  `text.replaceAll(' ', '\u{000A0}')`
We started to use it and it seems that it "just works" for all our use cases. It just replaces all spaces in text with Unicode [No-Break Space](https://unicode-table.com/en/00A0/) character.

Regarding Chinese characters, this workaround did not work too🥲

<details>
<summary> Sample Code </summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    const chineseCharacters = '漢字一漢字二漢字三漢字四漢字五漢字六漢字七漢字八';
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home Page'),
      ),
      body: Center(
        child: SizedBox(
          width: 300,
          child: Column(
            children: [
              const Text(
                'The following Chinese characters have white space at the end.',
              ),
              Container(
                color: Colors.green,
                child: const Text(
                  chineseCharacters,
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
              const Text('The workaround does not work.'),
              Container(
                color: Colors.green,
                child: Text(
                  chineseCharacters.replaceAll(' ', '\u{000A0}'),
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
              const Text('The following text shows \'...\' properly.'),
              Container(
                color: Colors.green,
                child: const Text(
                  '漢字一 漢字二 漢字三 漢字四 漢字五 漢字六漢字七',
                  overflow: TextOverflow.ellipsis,
                  maxLines: 1,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```


</details>


<details>
<summary> Result Image </summary>

<img src="https://user-images.githubusercontent.com/12263798/200107721-393b89c8-aecf-4a4e-9a46-1edb6778011a.png" width="320">


</details>

<details>
<summary>  flutter doctor -v </summary>

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 2.5.2, on macOS 12.6 21G115 darwin-arm, locale ja-JP)
    • Flutter version 2.5.2 at /Users/rls01036096/fvm/versions/2.5.2
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3595343e20 (1 year, 1 month ago), 2021-09-30 12:58:18 -0700
    • Engine revision 6ac856380f
    • Dart version 2.14.3

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/rls01036096/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/rls01036096/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode13.4.1.app/Contents/Developer
    • Xcode 13.4.1, Build version 13F100
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.72.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.52.0

[✓] Connected device (3 available)
    • iPhone 11 (mobile)                   • 12C240B7-596E-4563-BE11-7C533353F189 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-15-5 (simulator)
    • iPod touch (7th generation) (mobile) • DA40B225-2EBB-4A39-81BF-6A544B884DDF • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-15-5 (simulator)
    • Chrome (web)                         • chrome                               • web-javascript • Google Chrome 107.0.5304.87

• No issues found!

```

</details>

**Comment by mminhlequang** [comment]
![image](https://user-images.githubusercontent.com/46816270/209114444-65da107f-9d96-4737-849b-0ffec3a7685e.png)
I want ask about this, i used `overflow: TextOverflow.ellipsis,` with font `outfit` flutter web

**Comment by thailocnexle** [comment]
> > It is not a solution. It's a workaround. Be very cautious with using it. It changes how characters are displayed. Differences are subtle, but overall, letters are a bit wider.
> 
> Yes, this workaround increases letter spacing so it is not suitable for us. I just found another workaround which I not see in this thread:
> 
> ```dart
>  text.replaceAll(' ', '\u{000A0}')
> ```
> 
> We started to use it and it seems that it "just works" for all our use cases. It just replaces all spaces in text with Unicode [No-Break Space](https://unicode-table.com/en/00A0/) character.
> 
> Hope it may help someone :)

this workaround make some Arab character broken (الماس =>  ا  ل م ا س). need fix by Flutter team.

**Comment by Mahmoud-ElShenawy** [comment]
Use this extension 
```
extension StringEllipsis on String {
  String tight() {
    return Characters(this)
        .replaceAll(Characters(' '), Characters('\u{000A0}'))
        .toString();
  }
}
```
then use it as the following
```
Row(
         children: [
                              Flexible(
                                child: Text(
                                 value.tight(),
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                ),
                              ),
                         ],
 ),

```

Working fine with Arabic

**Comment by Lonyless** [comment]
Will this be fixed?

**Comment by mack-at-pieces** [comment]
A viable and just as elegant solution is to simply use `TextOverflow.fade` instead @RaymondBakker @Lonyless. That is what we've done for our products and its been just fine.

**Comment by dawid-niedzwiecki** [comment]
> > Will this be fixed?
> 
> No, this won't get fixed.

why not

**Comment by Lonyless** [comment]
@mack-at-pieces thanks for the tip, but unfortunately it wont for for my use case.

**Comment by Hixie** [comment]
Status update: Fixing this issue requires changes in Skia. We're working with the Skia team to get them an up to date list of issues people would like fixed, and this will be at the top of the list.

**Comment by feinstein** [comment]
Great to know! Will this be a problem in Impeller as well or is it already working there?

**Comment by tvolkert** [comment]
Great question @feinstein.  We still use Skia for text rendering and layout, so Impeller won't affect this.

**Comment by anthonymoretti** [comment]
> Yes, this workaround increases letter spacing so it is not suitable for us. I just found another workaround which I not see in this thread:
> 
> ```dart
>  text.replaceAll(' ', '\u{000A0}')
> ```
> 
> We started to use it and it seems that it "just works" for all our use cases. It just replaces all spaces in text with Unicode [No-Break Space](https://unicode-table.com/en/00A0/) character.
> 
> Hope it may help someone :)

@roman-petrov that's a good fix. If you make one more modification it also replaces hyphens with non-breaking hyphens:

```dart
'UFC 284 - Makhachev vs Volkanovski'
    .characters
    .replaceAll(Characters(' '), Characters('\u{00A0}'))
    .replaceAll(Characters('-'), Characters('\u{2011}'))
    .toString()
```

Without the second `replaceAll` the text looks like this:

"UFC 284 -..."

Instead of this:

"UFC 284 - Makhachev vs V..."

**Comment by fullflash** [comment]
my current workaround is to replace.

`str.replaceAll('', '\u200B')`

**Comment by anthonymoretti** [comment]
Hi @fullflash, if you do it that way just be wary of issues with letter spacing, emojis and hyphens.

**Comment by fullflash** [comment]
> ```dart
>  .characters
>     .replaceAll(Characters(' '), Characters('\u{00A0}'))
>     .replaceAll(Characters('-'), Characters('\u{2011}'))
>     .toString()
> ```

Yes but this solutions also results incorrect breaking chars like "u" letter in word "you" as you can see in the screenshot

![image](https://user-images.githubusercontent.com/19576508/228672077-969c3c2e-ccce-4f46-a0ba-32961827230e.png)

**Comment by anthonymoretti** [comment]
@fullflash You don't need to apply the replacements for multiline text fields.

**Comment by zanderso** [comment]
Closing as presumed fixed by https://skia.googlesource.com/skia.git/+log/0d31aa1f49cb8cfd790c44d92a9a023839524865

**Comment by MisterLight** [comment]
This fix was added in version 3.7.12? Looks like nothing changed

**Comment by POlczak-ITTouch** [comment]
@MisterLight The [changelog](https://github.com/flutter/flutter/wiki/Hotfixes-to-the-Stable-Channel) for 3.7.12 doesn't mention any changes related to this problem. The fix must be a part of some upcoming release

**Comment by zanderso** [comment]
This will not be backported to 3.7. I have requested a CP into the 3.10 beta here https://github.com/flutter/flutter/issues/125308.
